### PR TITLE
Stabilize the pattern overrides block context

### PR DESCRIPTION
**Pull request by talldan**
## What?
This PR stabilizes the pattern overrides block context, and partially removes the pattern overrides experiment.

Four core blocks are currently supported for pattern overrides, and the code in `lib/block-supports/pattern.php` previously added a way for those blocks to `useContext` when the pattern experiment was active. In this PR I'm instead adding the `useContext` declarations directly to the block.json files.

## Why?
This should help with:
- Backports to core - the code in the pattern.php file no longer needs to be backported, the blocks wil be automatically updated in core along with package updates.
- Removing the pattern overrides experiment flag - the flag will be removed as part of shipping the feature in WordPress 6.5. The feature can also be released in Gutenberg 17.6, though this will require also removing the block binding experiment.

## Testing Instructions
- Test patterns with the 'Pattern overrides' experiment turned off
- Test patterns with the 'Pattern overrides' experiment turned on ([see instructions here](https://github.com/WordPress/gutenberg/issues/53705#issuecomment-1882305331))

Both should work the same as in `trunk`.

### CODE REVIEW

**Review by kevin940726 — APPROVED** [review]
Tested well and code looks good! 👍

File: packages/block-library/src/button/block.json [hunk 1]
```diff
@@ -8,6 +8,7 @@
 	"description": "Prompt visitors to take action with a button-style link.",
 	"keywords": [ "link" ],
 	"textdomain": "default",
+	"usesContext": [ "pattern/overrides" ],
 	"attributes": {
```
Comment (kevin940726) [inline packages/block-library/src/button/block.json:12]: Should we add the `overrides` attribute here or should we do in #58105?

File: packages/block-library/src/block/index.php
Comment (kevin940726) [inline packages/block-library/src/block/index.php:1]: Should we add the `overrides` attribute in `block.json` or should we do in #58105?

Comment (talldan) [reply packages/block-library/src/block/index.php:1]: Oh yeah, I did make that change but reverted it accidentally, I'll bring it back again.

Comment (talldan) [reply packages/block-library/src/block/index.php:1]: Changed in [c8ca2e9](https://github.com/WordPress/gutenberg/pull/58102/commits/c8ca2e9de387440e9f57a49d256ec1cdf9f0021b)

**Comment by getdave** [comment]
@talldan I saw you noted that `patterns.php` has been removed. So we're good on this PR in terms of backports for 6.5?

**Comment by talldan** [comment]
Yep, I think we managed to move all the logic into block code, apart from the registration using the block binding API, but there's the block bindings backport PR that handles that.

**Comment by getdave** [comment]
✅ I updated the [PHP Sync Tracking Issue for WP 6.5](https://github.com/WordPress/gutenberg/issues/57959) to note this PR does not require a backport for WP 6.5.

**Pull request merged**